### PR TITLE
Solc upgrade

### PR DESCRIPTION
**Pull request by iamchrissmith**

### CODE REVIEW

File: src/plans/D3MAaveDaiPlan.sol [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.8.13;
```
Comment (naszam) [inline src/plans/D3MAaveDaiPlan.sol:17]: wondering if this should be `>=0.6.12` and deployed with 0.8.13 instead, as I think it could break compilation as dependency of `dss-exec-lib` in tests (`DssAction.t.sol`) for the `bar` action.
cc: @brianmcmichael

File: Makefile [hunk 2]
```diff
@@ -2,5 +2,5 @@ all    		 :; DAPP_BUILD_OPTIMIZE=1 DAPP_BUILD_OPTIMIZE_RUNS=200 dapp --use solc:
 clean  		 :; dapp clean
 test   		 :; ./test.sh match=$(match) optimizer=1
 test-dev   	 :; ./test.sh match=$(match) optimizer=0
```
Comment (naszam) [inline Makefile:4]: same as above

Comment (iamchrissmith) [reply Makefile:4]: you mean adding quotes?

Comment (naszam) [reply Makefile:4]:
```suggestion
test   		 :; ./test.sh match="$(match)" optimizer=1
test-dev   	 :; ./test.sh match="$(match)" optimizer=0
```

File: src/DssDirectDepositHub.sol [hunk 3]
```diff
@@ -298,7 +288,7 @@ contract DssDirectDepositHub {
         if (assetBalance > daiDebt) {
             fees = assetBalance - daiDebt;
```
Comment (iamchrissmith) [inline src/DssDirectDepositHub.sol:289]: Use `unchecked { ... }` here and in other places where we don't need safemath

Comment (brianmcmichael) [reply src/DssDirectDepositHub.sol:289]: Can't comment on the line since it's not updated, but in the Hub exec function, can also do:

`toUnwind = Art - lineWad;` as unchecked I think.

File: src/plans/D3MAaveDaiPlan.sol (hunk 1 above)
Comment (brianmcmichael) [reply src/plans/D3MAaveDaiPlan.sol]: I definitely don't want to hinder progress toward the later solidity version on behalf of dss-exec-lib. We can probably use the rpc tests for this for now until we have a way to manage different solc versions.

File: src/plans/D3MAaveDaiPlan.sol [hunk 4]
```diff
@@ -154,19 +145,16 @@ contract D3MAaveDaiPlan is ID3MPlan {
         // Do inverse calculation of interestStrategy
         uint256 variableRateSlope1 = interestStrategy.variableRateSlope1();
         uint256 targetUtil;
-        if (targetInterestRate > _add(base, variableRateSlope1)) {
+        if (targetInterestRate > base + variableRateSlope1) {
             // Excess interest rate
             uint256 r = targetInterestRate - base - variableRateSlope1;
```
Comment (iamchrissmith) [inline src/plans/D3MAaveDaiPlan.sol:150]: unchecked

File: src/plans/D3MAaveDaiPlan.sol [hunk 5]
```diff
@@ -154,19 +145,16 @@ contract D3MAaveDaiPlan is ID3MPlan {
...
-                            interestStrategy.OPTIMAL_UTILIZATION_RATE()
-                        );
+            targetUtil = _rdiv(
+                            _rmul(interestStrategy.EXCESS_UTILIZATION_RATE(), r),
+                            interestStrategy.variableRateSlope2()
+                        ) + interestStrategy.OPTIMAL_UTILIZATION_RATE();
         } else {
             // Optimal interest rate
-            targetUtil = _rdiv(_rmul(_sub(targetInterestRate, base), interestStrategy.OPTIMAL_UTILIZATION_RATE()), variableRateSlope1);
+            targetUtil = _rdiv(_rmul((targetInterestRate - base), interestStrategy.OPTIMAL_UTILIZATION_RATE()), variableRateSlope1);
```
Comment (iamchrissmith) [inline src/plans/D3MAaveDaiPlan.sol:157]: refactor to be multiline

File: Makefile [hunk 6]
```diff
@@ -1,6 +1,6 @@
 all    		 :; DAPP_BUILD_OPTIMIZE=1 DAPP_BUILD_OPTIMIZE_RUNS=200 dapp --use solc:0.6.12 build
 clean  		 :; dapp clean
-test   		 :; ./test.sh match=$(match) optimizer=1
-test-dev   	 :; ./test.sh match=$(match) optimizer=0
-test-forge 	 :; ./test-forge.sh match=$(match) block=$(block) match-test=$(match-test) match-contract=$(match-contract)
+test   		 :; ./test.sh match="$(match)" optimizer=1
+test-dev   	 :; ./test.sh match="$(match)" optimizer=0
+test-forge 	 :; ./test-forge.sh match="$(match)" block="$(block)" match-test="$(match-test)" match-contract="$(match-contract)"
 deploy 		 :; make && dapp create DssDirectDepositAaveDai 0xdA0Ab1e0017DEbCd72Be8599041a2aa3bA7e740F 0x4449524543542d4141564556322d444149000000000000000000000000000000 0x7d2768dE32b0b80b7a3454c06BdAc94A69DDc7A9 0xd784927Ff2f95ba542BfC824c8a8a98F3495f6b5
```
Comment (gbalabasquer) [inline Makefile:6]: Not the correct PR but I guess as we are changing other stuff here we could remove this `deploy` command which is old.

File: src/DssDirectDepositHub.sol [hunk 7]
```diff
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-pragma solidity 0.6.12;
+pragma solidity >=0.6.12;
```
Comment (gbalabasquer) [inline src/DssDirectDepositHub.sol:17]: Keeping the pragma this way is actually a bug. As compiling with solc 0.6.x will make the safe math is not applied and then the code has a complete different behaviour. Not sure if 0.7.0 is already implementing it but to be honest I'd just do:
```
pragma solidity ^0.8.13;
```

File: src/DssDirectDepositHub.sol [hunk 8]
```diff
@@ -530,7 +519,7 @@ contract DssDirectDepositHub {
         address vow_ = vow;
         uint256 wad = vat.gem(ilk, address(pool));
         require(wad < 2 ** 255, "DssDirectDepositHub/overflow");
-        vat.suck(vow_, vow_, _mul(wad, RAY)); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
+        vat.suck(vow_, vow_, (wad * RAY)); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
```
Comment (gbalabasquer) [inline src/DssDirectDepositHub.sol:522]:
```suggestion
        vat.suck(vow_, vow_, wad * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
```

File: src/DssDirectDepositHub.sol [hunk 9]
```diff
@@ -406,7 +396,6 @@ contract DssDirectDepositHub {
                     currentAssets
                 );
             } else {
-                // All the subtractions are safe as otherwise toUnwind is > 0
                 uint256 toWind = targetAssets - currentAssets;
                 toWind = _min(toWind, lineWad - Art);
                 toWind = _min(toWind, (Line - debt) / RAY);
```
Comment (gbalabasquer) [inline src/DssDirectDepositHub.sol:401]: As Brian pointed out above, this can be inside `unchecked`

File: src/plans/D3MAaveDaiPlan.sol [hunk 10]
```diff
@@ -175,17 +163,14 @@ contract D3MAaveDaiPlan is ID3MPlan {
...
-                TokenLike(dai).balanceOf(adai),
-                totalDebt
-            );
+        uint256 totalPoolSize = TokenLike(dai).balanceOf(adai) + totalDebt;
 
         uint256 targetTotalPoolSize = _calculateTargetSupply(targetInterestRate, totalDebt);
         if (targetTotalPoolSize >= totalPoolSize) {
             // Increase debt (or same)
-            return _add(currentAssets, targetTotalPoolSize - totalPoolSize);
+            return currentAssets + (targetTotalPoolSize - totalPoolSize);
```
Comment (gbalabasquer) [inline src/plans/D3MAaveDaiPlan.sol:173]: in theory we could separate the math here to make the subtraction unchecked but doesn't seem to be worth it.

File: src/DssDirectDepositHub.sol (hunk 7 above)
Comment (hexonaut) [reply src/DssDirectDepositHub.sol:17]: Yes strongly agree with @gbalabasquer here.

File: Makefile [hunk 11]
```diff
@@ -1,6 +1,6 @@
 all    		 :; DAPP_BUILD_OPTIMIZE=1 DAPP_BUILD_OPTIMIZE_RUNS=200 dapp --use solc:0.6.12 build
```
Comment (talbaneth) [inline Makefile:1]:
```suggestion
all    		 :; DAPP_BUILD_OPTIMIZE=1 DAPP_BUILD_OPTIMIZE_RUNS=200 dapp --use solc:0.8.13 build
```

File: src/plans/D3MAaveDaiPlan.sol [hunk 12]
```diff
@@ -154,19 +145,16 @@ contract D3MAaveDaiPlan is ID3MPlan {
...
-                            _rdiv(
-                                _rmul(interestStrategy.EXCESS_UTILIZATION_RATE(), r),
-                                interestStrategy.variableRateSlope2()
-                            ),
-                            interestStrategy.OPTIMAL_UTILIZATION_RATE()
-                        );
+            targetUtil = _rdiv(
+                            _rmul(interestStrategy.EXCESS_UTILIZATION_RATE(), r),
+                            interestStrategy.variableRateSlope2()
+                        ) + interestStrategy.OPTIMAL_UTILIZATION_RATE();
```
Comment (talbaneth) [inline src/plans/D3MAaveDaiPlan.sol:154]: Indentation changed here to 3 spaces instead of 4.
Also the closing bracket of the `_rdiv` is not completely aligned to it (missing one space before the closing bracket).

File: src/plans/D3MAaveDaiPlan.sol (hunk 5 above)
Comment (talbaneth) [reply src/plans/D3MAaveDaiPlan.sol:157]: Also consider removing the brackets around `targetInterestRate - base`

File: src/plans/D3MAaveDaiPlan.sol [hunk 13]
```diff
@@ -175,17 +163,14 @@ contract D3MAaveDaiPlan is ID3MPlan {
...
         if (targetInterestRate == 0) return 0;  // De-activated
 
-        uint256 totalDebt = _add(stableDebt.totalSupply(), variableDebt.totalSupply());
+        uint256 totalDebt = stableDebt.totalSupply() + variableDebt.totalSupply();
 
-        uint256 totalPoolSize = _add(
-                TokenLike(dai).balanceOf(adai),
-                totalDebt
-            );
+        uint256 totalPoolSize = TokenLike(dai).balanceOf(adai) + totalDebt;
```
Comment (talbaneth) [inline src/plans/D3MAaveDaiPlan.sol:168]:
```suggestion
        uint256 totalPoolSize = dai.balanceOf(adai) + totalDebt;
```

File: src/plans/D3MAaveDaiPlan.sol (hunk 1 above)
Comment (gbalabasquer) [reply src/plans/D3MAaveDaiPlan.sol]: I hadn't seen this comment before, but as I mentioned in mine, we must not leave a pragma that allows to compile the contract in a broken state and that's what means leaving the `>=0.6.12` one here.
`dss-exec-lib` will need to figure it out how to include this dependency, might be better just a mock or worst case maintain a branch of this repo for it.

File: Makefile (hunk 6 above)
Comment (iamchrissmith) [reply Makefile:6]: i'd prefer to keep this out of this PR.  probably belongs in v2

File: shell.nix [hunk 14]
```diff
@@ -1,11 +1,11 @@
 { url
   , dappPkgs ? (
     import (fetchTarball "https://github.com/makerdao/makerpkgs/tarball/master") {}
-  ).dappPkgsVersions.hevm-0_43_1
+  ).dappPkgsVersions.master-20220406
```
Comment (gbalabasquer) [inline shell.nix:4]:
```suggestion
  ).dappPkgsVersions.master-20220524
```

File: shell.nix [hunk 15]
```diff
@@ -1,11 +1,11 @@
...
 }: with dappPkgs;
 
 mkShell {
-  DAPP_SOLC = solc-static-versions.solc_0_6_12 + "/bin/solc-0.6.12";
+  DAPP_SOLC = solc-static-versions.solc_0_8_13 + "/bin/solc-0.8.13";
```
Comment (gbalabasquer) [inline shell.nix:8]:
```suggestion
  DAPP_SOLC = solc-static-versions.solc_0_8_14 + "/bin/solc-0.8.14";
```

File: src/DssDirectDepositHub.sol [hunk 16]
```diff
@@ -223,7 +213,7 @@ contract DssDirectDepositHub {
     }
 
     /**
-        @notice remove nope on daiJoin to kill hub
+        @notice remove hope on daiJoin to kill hub
```
Comment (gbalabasquer) [inline src/DssDirectDepositHub.sol:216]: We might want to revert this change and evaluate this function in another PR. I think this is useless. Removing the `vat` access to the `daiJoin` should not be really needed and we should avoid adding non useful functions.

File: src/DssDirectDepositHub.sol [hunk 17]
```diff
@@ -378,16 +371,22 @@ contract DssDirectDepositHub {
...
             uint256 toUnwind;
             if (Art > lineWad) {
-                toUnwind = Art - lineWad;
+                unchecked {
+                    toUnwind = Art - lineWad;
+                }
             }
             if (debt > Line) {
-                toUnwind = _max(toUnwind, _divup(debt - Line, RAY));
+                unchecked {
```
Comment (gbalabasquer) [inline src/DssDirectDepositHub.sol:379]: Was wondering if the `unchecked` might modify the behaviour of `_divup`, but no, nothing to worry about:
"The setting only affects the statements that are syntactically inside the block. Functions called from within an unchecked block do not inherit the property."
https://docs.soliditylang.org/en/v0.8.14/control-structures.html

File: src/DssDirectDepositHub.sol [hunk 18]
```diff
@@ -306,7 +299,7 @@ contract DssDirectDepositHub {
         require(amount <= 2 ** 255, "DssDirectDepositHub/overflow");
```
[the remark this answers is not in the document]
Comment (gbalabasquer) [reply src/DssDirectDepositHub.sol:299]: Just to clarify, the net result is the same for solc 0.8.x as will revert anyway, I'm proposing it reverts in the line that actually checks the overflow than later.
If we would like to replicate the exact same behaviour than with solc < 0.8.0, we should complicate the code and is not worth it for that extra wei.
I know for this specific function we are probably moving to the extra value on `min` but that can be addressed later in another PR from this better meaning of the `require`.

[the remark this answers is not in the document]
Comment (gbalabasquer) [reply src/DssDirectDepositHub.sol:299]: On `uncull` we should also change the format of the require.

File: test.sh [hunk 19]
```diff
@@ -10,16 +10,16 @@ do
...
+            *)
     esac
 done
 
 export DAPP_BUILD_OPTIMIZE="$OPTIMIZER"
 export DAPP_BUILD_OPTIMIZE_RUNS=200
 
 if [[ -z "$1" ]]; then
-  dapp --use solc:0.6.12 test --rpc-url="$ETH_RPC_URL" -v
+  dapp --use solc:0.8.13 test --rpc-url="$ETH_RPC_URL" -v
```
Comment (gbalabasquer) [inline test.sh:22]:
```suggestion
  dapp --use solc:0.8.14 test --rpc-url="$ETH_RPC_URL" -v
```

File: test.sh [hunk 20]
```diff
@@ -10,16 +10,16 @@ do
...
 else
-  dapp --use solc:0.6.12 test --rpc-url="$ETH_RPC_URL" --match "$MATCH" -vv
+  dapp --use solc:0.8.13 test --rpc-url="$ETH_RPC_URL" --match "$MATCH" -vv
```
Comment (gbalabasquer) [inline test.sh:24]:
```suggestion
  dapp --use solc:0.8.14 test --rpc-url="$ETH_RPC_URL" --match "$MATCH" -vv
```

File: src/DssDirectDepositHub.sol (hunk 18 above)
[the remark this answers is not in the document]
Comment (gbalabasquer) [reply src/DssDirectDepositHub.sol:299]: I think `require(amount <= type(int256).max, "DssDirectDepositHub/overflow");` is the best format.

**Comment by gbalabasquer** [comment]
@iamchrissmith you might want to merge https://github.com/makerdao/dss-direct-deposit/pull/49 to v2 and then v2 to this before merging this PR.

File: src/DssDirectDepositHub.sol (hunk 18 above)
[the remark this answers is not in the document]
Comment (iamchrissmith) [reply src/DssDirectDepositHub.sol]: done https://github.com/makerdao/dss-direct-deposit/pull/47/commits/70e46b4ede0ec98ac6949e2d780c3ed9b5cb7b17

**Review by gbalabasquer — APPROVED** [review]
LGTM

File: src/DssDirectDepositHub.sol [hunk 21]
```diff
@@ -442,10 +441,12 @@ contract DssDirectDepositHub {
                     currentAssets
                 );
             } else {
-                // All the subtractions are safe as otherwise toUnwind is > 0
```
Comment (naszam) [inline src/DssDirectDepositHub.sol:445]: why it was removed?

Comment (iamchrissmith) [reply src/DssDirectDepositHub.sol:445]: i think i removed it before `unchecked` was added so safemath was in place.  adding it back

**Review by gbalabasquer — APPROVED** [review]
Renewing approval.

File: src/plans/D3MPlanBase.t.sol [hunk 22]
```diff
@@ -88,7 +88,7 @@ contract D3MPlanBaseTest is DSTest {
             address(bytes20(uint160(uint256(keccak256("hevm cheat code")))))
         );
 
-        dai = DaiLike(123);
+        dai = DaiLike(address(123));
```
Comment (naszam) [inline src/plans/D3MPlanBase.t.sol:91]:
```suggestion
        dai = DaiLike(address(0x123));
```

File: src/pools/D3MAaveDaiPool.sol [hunk 23]
```diff
@@ -158,7 +151,7 @@ contract D3MAaveDaiPool is ID3MPool {
         // Verify the correct amount of adai shows up
         uint256 interestIndex = pool.getReserveNormalizedIncome(address(asset));
         uint256 scaledAmount = _rdiv(wad, interestIndex);
-        return adai.scaledBalanceOf(address(this)) >= _add(scaledPrev, scaledAmount);
+        return adai.scaledBalanceOf(address(this)) >= scaledPrev + scaledAmount;
```
Comment (naszam) [inline src/pools/D3MAaveDaiPool.sol:154]:
```suggestion
        return adai.scaledBalanceOf(address(this)) >= (scaledPrev + scaledAmount);
```

File: src/plans/D3MPlanBase.t.sol [hunk 24]
```diff
@@ -88,7 +88,7 @@ contract D3MPlanBaseTest is DSTest {
             address(bytes20(uint160(uint256(keccak256("hevm cheat code")))))
         );
 
-        dai = DaiLike(address(123));
+        dai = DaiLike(address(0x123));
```
Comment (gbalabasquer) [inline src/plans/D3MPlanBase.t.sol:91]: It's pretty common we do `address(123)` in general, it just means something random.
I mean not against to the change, but don't think it is a fix at all.

File: src/pools/D3MAaveDaiPool.sol [hunk 25]
```diff
@@ -151,7 +151,7 @@ contract D3MAaveDaiPool is ID3MPool {
         // Verify the correct amount of adai shows up
         uint256 interestIndex = pool.getReserveNormalizedIncome(address(asset));
         uint256 scaledAmount = _rdiv(wad, interestIndex);
-        return adai.scaledBalanceOf(address(this)) >= scaledPrev + scaledAmount;
+        return adai.scaledBalanceOf(address(this)) >= (scaledPrev + scaledAmount);
```
[the remark this answers is not in the document]
Comment (gbalabasquer) [reply src/pools/D3MAaveDaiPool.sol:154]: The sum happens before the comparison.

[the remark this answers is not in the document]
Comment (gbalabasquer) [reply src/pools/D3MAaveDaiPool.sol:154]: I think we even removed the parenthesis from same places where wasn't necessary (not every place though).

[the remark this answers is not in the document]
Comment (iamchrissmith) [reply src/pools/D3MAaveDaiPool.sol:154]: i understood @naszam 's comment to suggest that the previous method could have been evaluated as:
1. is scaledBalanceOf >= scaledPrev -> 0 or 1 (bool)
2. add scaledAmount to 0 or 1

So this ensures we do the math first.  0.8 did change some order of operations things, but even if this wasn't a bug, i think the `(...)` is a little clearer and doesn't cost anything (right?)

[the remark this answers is not in the document]
Comment (naszam) [reply src/pools/D3MAaveDaiPool.sol:154]: I left a suggestion for consistency with the previous function `_add` that was returning the sum, I would consider it a nit, I should have clarified that.

File: src/plans/D3MPlanBase.t.sol (hunk 24 above)
Comment (naszam) [reply src/plans/D3MPlanBase.t.sol:91]: yeah, same as above, was meant to be a nit

File: src/pools/D3MAaveDaiPool.sol (hunk 25 above)
[the remark this answers is not in the document]
Comment (naszam) [reply src/pools/D3MAaveDaiPool.sol:154]: I don't have a strong opinion on this, if without parenthesis is cleaner I'm ok with that.

[the remark this answers is not in the document]
Comment (iamchrissmith) [reply src/pools/D3MAaveDaiPool.sol:154]: @gbalabasquer also do not feel strongly here.  You pick with `(...)` or without

File: src/plans/D3MPlanBase.t.sol (hunk 24 above)
Comment (iamchrissmith) [reply src/plans/D3MPlanBase.t.sol:91]: i'd revert then for consistency with our other dummy addresses

Comment (iamchrissmith) [reply src/plans/D3MPlanBase.t.sol]: e76057a

**Review by naszam — APPROVED** [review]
LGTM
